### PR TITLE
Fix flow bugs

### PR DESCRIPTION
**Pull request by DanielHeath**

### CODE REVIEW

File: script/typecheck [hunk 1]
```diff
@@ -11,6 +11,24 @@ function check {
     echo "Good"
 }
 
+badFlow=$( grep '@flow' -n -r src test | grep -v :1: | sed s/:.*// || echo "" )
```
Comment (danni) [inline script/typecheck]: Use `read -n 1`

**Comment by DanielHeath** [comment]
Opened https://github.com/facebook/flow/issues/902

File: script/typecheck [hunk 2]
```diff
@@ -11,6 +11,26 @@ function check {
     echo "Good"
 }
 
+echo "${BOLD}Checking for missing @flow comments${END}"
+echo "Good"
+
+missingflow=""
```
Comment (danni) [inline script/typecheck]: you can literally use the binaries true/false.

So `HAVE_FLOW=true`

File: script/typecheck [hunk 3]
```diff
@@ -11,6 +11,26 @@ function check {
...
+for file in $( git ls-files '**/*.js' ); do
+    read flowtype < "$file"
+    if test "x$flowtype" = 'x/* flow:disable */'; then
+        echo "Skipping disabled file $file" > /dev/null
+    elif test "x$flowtype" != 'x/* @flow */'; then
```
Comment (danni) [inline script/typecheck:22]: `HAVE_FLOW=HAVE_FLOW && test ...`

File: script/typecheck [hunk 4]
```diff
@@ -11,6 +11,26 @@ function check {
...
+        missingflow="true"
+        cat << EOF
+$file is not annotated for flowtype (must be the first line)! - got "$flowtype"
+EOF
+    fi
+done
+
+if [ "$missingflow" = "true" ] ; then
```
Comment (danni) [inline script/typecheck]: `if ! HAVE_FLOW; then`

File: src/iss.js [hunk 5]
```diff
@@ -128,9 +129,13 @@ export function mungeUrlQuery(url_: string, data: Object): string {
     return url_ + joiner + serialized;
 }
 
-export async function request(path: string, data: ?searchRequest): Object {
+export async function request(
+    path: string,
+    data: ?searchRequest
+): Promise<Object> {
```
Comment (danni) [inline src/iss.js:135]: I think now `async` functions no longer need to return `Promise<blah>` and can just return `blah`.

File: src/iss.js [hunk 6]
```diff
@@ -128,9 +129,13 @@ export function mungeUrlQuery(url_: string, data: Object): string {
...
     var url_: string = ISS_URL || process.env.ISS_URL;
 
+    /* flow:disable url module does something truly bizarre to flow */
```
Comment (danni) [inline src/iss.js]: ... Is it that we're redefining `url` somewhere? That's typically the problem.

File: src/iss.js [hunk 7]
```diff
@@ -295,6 +300,7 @@ export class Service {
         }
 
         var request_: searchRequest = {
+            q: "",
```
Comment (danni) [inline src/iss.js]: Not correct.

Comment (danni) [inline src/iss.js]: You want to define `q` as optional. Sending `q` in the query will change what happens.

File: src/storage.js [hunk 8]
```diff
@@ -15,16 +15,23 @@ class Storage {
         sessionstorage.setItem(key, obj);
     }
 
-    static getJSON(key: string): ?(Object|Array) {
+    static getJSON(key: string): any {
```
Comment (danni) [inline src/storage.js:18]: Why?

File: src/storage.js [hunk 9]
```diff
@@ -15,16 +15,23 @@ class Storage {
...
-            return JSON.parse(this.getItem(key));
+            return JSON.parse(item);
         } catch (error) {
             console.error(error);
             return null;
         }
     }
 
-    static setJSON(key: string, obj: Object|Array): void {
+    static setJSON(key: string, obj: mixed): void {
```
Comment (danni) [inline src/storage.js:34]: Why?

File: webpack/utils/write-stats.js [hunk 10]
```diff
@@ -26,8 +28,7 @@ module.exports = function writeStats(stats) {
         var unorderedChunks = _(allChunkNames).difference(chunkOrdering);
         var sortedChunkNames = chunkOrdering.concat(unorderedChunks);
 
-        for (var name of sortedChunkNames) {
-
+        for (var name of sortedChunkNames) if (json.assetsByChunkName[name]) {
```
Comment (danni) [inline webpack/utils/write-stats.js]: One block per line please. This sort of construct just leads to mistakes.

**Comment by danni** [comment]
So if Flow now supports `let` and `const`, should we switch to using let/const. Since I think I've specified in eslint to apply `let` semantics to all uses of `var`, it should be easy to `sed` them all to `let` and swap eslint over to denying `var` and requiring `let`/`const`. We can then go back to adding `const` where appropriate.

**Comment by danni** [comment]
Using more `const` would probably also help with things like where we accidentally override the URL module with a variable called URL. Although I would expect there's already a eslint rule for shadow declarations we could turn on.

**Comment by DanielHeath** [comment]
That sounds good to me, but I'm keenly aware of how fast this PR bitrots (any new file in master causes it to fail the merge build). Happy to do that separately?

File: script/typecheck (hunk 3 above)
Comment (DanielHeath) [inline script/typecheck:22]: assignment inside an `if`?

Comment (danni) [inline script/typecheck:22]: Oh, I missed the second line. Then you can just do `HAVE_FLOW=false`.

File: src/storage.js (hunk 8 above)
Comment (DanielHeath) [inline src/storage.js:18]: `JSON.parse` is typed `any` in flow - it's valid to do `JSON.parse('"a string"')`

File: src/storage.js (hunk 9 above)
Comment (DanielHeath) [inline src/storage.js:34]: The union type doesn't work in function declarations but this file had never been typechecked.

File: src/iss.js (hunk 6 above)
Comment (DanielHeath) [inline src/iss.js]: I've filed a bug with flow - `url.format(url.parse('a url'))` doesn't work.

File: src/storage.js (hunk 9 above)
Comment (danni) [inline src/storage.js:34]: I don't think that's correct. According to the docs:

```
Sometimes, multiple signatures are not needed to express overloading: the signatures can 
be coalesced using union types. Flow provides the following syntax for union types:

T1 | .. | Tn

is the union of types Ti. Union types are available for general use.
```

File: src/iss.js (hunk 6 above)
Comment (danni) [inline src/iss.js]: URL in the comment?

File: src/storage.js (hunk 9 above)
Comment (DanielHeath) [inline src/storage.js:34]: As discussed: If you can get it working I'll be super happy, but I've sunk a significant bit of time into it without any progress.

**Comment by DanielHeath** [comment]
Failed due to hitting the saucelabs tunnel limit; other builds passed.

**Comment by DanielHeath** [comment]
Previous build passed, and latest commit is only a comment change (and passed the typecheck).

**Pull request merged**